### PR TITLE
Derive some traits

### PR DESCRIPTION
**Pull request by ratzrattillo**
It would be nice, if for Seify's types, traits like (Debug, Clone, PartialEq , Serialize, Deserialize) could be derived or implemented, where derivation is not possible.

### CODE REVIEW

File: src/range.rs [hunk 1]
```diff
@@ -10,7 +13,7 @@ pub enum RangeItem {
 }
 
 /// Range of possible values, comprised of individual values and/or intervals.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Serialize, Deserialize)] //Serialize might be risky due to large memory allocations with Soapy
```
Comment (bastibl) [inline src/range.rs:16]: Can you explain what this comment means?

File: src/args.rs [hunk 2]
```diff
@@ -64,7 +64,7 @@ impl Args {
     pub fn iter_mut(&mut self) -> std::collections::hash_map::IterMut<'_, String, String> {
         self.map.iter_mut()
     }
-    /// Get a reference to the underlying [HashMap](std::collections::HashMap).
+    /// Get a reference to the underlying [HashMap](HashMap).
```
Comment (bastibl) [inline src/args.rs:67]: If this works, we should be able to also remove the `(HashMap)`, no?

**Review by bastibl — COMMENTED** [review]
Thanks for the PR! I left some comments and also wondered whether you have some particular application or use-case in mind to derive `Serialize` and `Deserialize`. Are these all structs that required the trait for your application or is this still work-in-progress?

File: src/range.rs (hunk 1 above)
Comment (ratzrattillo) [reply src/range.rs:16]: I noticed that these large memory allocations have nothing to do with the serialization, but it seems to be a bladeRF issue. When calling `dev.frequency_range(Rx, 0).unwrap()` somehow a large amount of memory is allocated until the RAM is full and the application crashes. This does not happen with an RTL-SDR stick. Thus i would remove this comment again.

**Comment by ratzrattillo** [comment]
I thought about writing a web application to configure an SDR's Parameters. Something like this:
```rust
use std::fmt::Debug;
use std::sync::{Arc, Mutex};
use futuresdr::seify::{Args, Device, Direction::{self, Rx, Tx}, enumerate_with_args, Error, GenericDevice, Range};
use axum::{http::StatusCode, Json};
use axum::extract::State;
use axum::response::{IntoResponse, Response};
use serde::{Deserialize, Serialize};

#[derive(Clone, Default)]
pub struct DeviceHandles {
    pub rx_handle: Option<Device<GenericDevice>>,
    rx_attributes: Option<DeviceSelection>,
    pub tx_handle: Option<Device<GenericDevice>>,
    tx_attributes: Option<DeviceSelection>,
}

#[derive(Debug, Clone, PartialEq , Serialize, Deserialize)]
pub struct DeviceSelection {
    args: Args,
    direction: Direction,
    channel: usize,
}

#[derive(Debug, Serialize)]
struct ChannelProperties {
    frequency: Option<f64>,
    sample_rate: Option<f64>,
    gain: Option<f64>,
    frequency_range: Option<Range>,
    sample_rate_range: Option<Range>,
    gain_range: Option<Range>,
}

#[derive(Debug, Clone, Serialize)]
struct DeviceInfo<'a> {
    args: &'a Args,
    error: Option<Error>,
    num_rx_channels: Option<usize>,
    num_tx_channels: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelConfig {
    direction: Direction,
    channel: usize,
    value: f64,
}

enum Property {
    Frequency,
    SampleRate,
    Gain,
}

async fn set_property(property: Property, devices: Arc<Mutex<DeviceHandles>>, body: &ChannelConfig) -> Response {
    match devices.try_lock() {
        Ok(locked_devices) => {
            let dev = if body.direction == Rx {
                locked_devices.rx_handle.as_ref().unwrap().clone()
            } else {
                locked_devices.tx_handle.as_ref().unwrap().clone()
            };

            let res = match property {
                Property::Frequency => dev.set_frequency(body.direction, body.channel, body.value),
                Property::SampleRate => dev.set_sample_rate(body.direction, body.channel, body.value),
                Property::Gain => dev.set_gain(body.direction, body.channel, body.value),
            };

            match res {
                Ok(_) => { StatusCode::OK.into_response() }
                Err(err) => {
                    (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err.to_string())).into_response()
                }
            }
        }
        Err(err) => {
            (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err.to_string())).into_response()
        }
    }
}

async fn get_property(property: Property, devices: Arc<Mutex<DeviceHandles>>, body: &ChannelConfig) -> Response {
    match devices.try_lock() {
        Ok(locked_devices) => {
            let dev = if body.direction == Rx {
                locked_devices.rx_handle.as_ref().unwrap().clone()
            } else {
                locked_devices.tx_handle.as_ref().unwrap().clone()
            };

            let res = match property {
                Property::Frequency => dev.frequency(body.direction, body.channel),
                Property::SampleRate => dev.sample_rate(body.direction, body.channel),
                Property::Gain => dev.gain(body.direction, body.channel).transpose().unwrap(),
            };

            match res {
                Ok(value) => {
                    (StatusCode::OK, Json::from(ChannelConfig {
                        value,
                        direction: body.direction,
                        channel: body.channel,
                    })).into_response()
                }
                Err(err) => {
                    (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err.to_string())).into_response()
                }
            }
        }
        Err(err) => {
            (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err.to_string())).into_response()
        }
    }
}


//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn list_devices(
    Json(filter): Json<Args>,
) -> Response {
    //println!("filter: {:?}", filter);
    match enumerate_with_args(filter) {
        Ok(args_vec) => {
            //println!("args_vec: {:?}", args_vec);
            let device_infos: Vec<_> = args_vec
                .iter()
                .filter(|args| args.get::<String>("soapy_driver").unwrap_or_default() != "audio")// never allow audio devices
                .map(|args| Device::from_args(args).map_or_else(|err| {
                    DeviceInfo {
                        args,
                        error: Some(err),
                        num_rx_channels: None,
                        num_tx_channels: None,
                    }
                }, |dev| {
                    DeviceInfo {
                        args,
                        error: None,
                        num_rx_channels: Some(dev.num_channels(Rx).unwrap_or_default()),
                        num_tx_channels: Some(dev.num_channels(Tx).unwrap_or_default()),
                    }
                }))
                .collect();

            //println!("{:?}", device_infos);
            (StatusCode::OK, Json::from(device_infos)).into_response()
        }
        Err(err) => {
            (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err)).into_response()
        }
    }
}

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn select_device(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(device_selection): Json<DeviceSelection>, //serde_json::Value //
) -> Response { // https://docs.rs/axum/latest/axum/response/index.html#returning-different-response-types
    match Device::from_args(&device_selection.args) {
        Ok(dev) => {
            let channel_properties = ChannelProperties { // Other strategy would be using unwrap_or_default instead of Options
                frequency: dev.frequency(device_selection.direction, device_selection.channel).ok(),
                sample_rate: dev.sample_rate(device_selection.direction, device_selection.channel).ok(),
                gain: dev.gain(device_selection.direction, device_selection.channel).transpose().unwrap_or(Ok(0.)).ok(), // ok().unwrap_or_default() If we dont want null here, we have to fix that in the UI
                frequency_range: dev.frequency_range(device_selection.direction, device_selection.channel).ok(),
                sample_rate_range: dev.get_sample_rate_range(device_selection.direction, device_selection.channel).ok(),
                gain_range: dev.gain_range(device_selection.direction, device_selection.channel).ok(),
            };
            println!("{:?}", channel_properties);

            match devices.try_lock() {
                Ok(mut locked_devices) => {
                    if device_selection.direction == Rx {
                        locked_devices.rx_handle = Some(dev.clone());
                        locked_devices.rx_attributes = Some(device_selection.clone());
                    } else { // If its not Rx, it is Tx
                        locked_devices.tx_handle = Some(dev.clone());
                        locked_devices.tx_attributes = Some(device_selection.clone());
                    }

                    (StatusCode::OK, Json::from(channel_properties)).into_response()
                }
                Err(err) => {
                    (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err.to_string())).into_response()
                }
            }
        }
        Err(err) => {
            (StatusCode::INTERNAL_SERVER_ERROR, Json::from(err)).into_response()
        }
    }
}

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn set_frequency(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(body): Json<ChannelConfig>,
) -> Response {
    set_property(Property::Frequency, devices, &body).await
}

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn get_frequency(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(body): Json<ChannelConfig>,
) -> Response { get_property(Property::Frequency, devices, &body).await }

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn set_sample_rate(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(body): Json<ChannelConfig>,
) -> Response {
    set_property(Property::SampleRate, devices, &body).await
}

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn get_sample_rate(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(body): Json<ChannelConfig>,
) -> Response { get_property(Property::SampleRate, devices, &body).await }

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn set_gain(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(body): Json<ChannelConfig>,
) -> Response {
    set_property(Property::Gain, devices, &body).await
}

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn get_gain(
    State(devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(body): Json<ChannelConfig>,
) -> Response { get_property(Property::Gain, devices, &body).await }

//https://docs.rs/axum/latest/axum/extract/index.html
#[axum_macros::debug_handler]
pub async fn start(
    State(_devices): State<Arc<Mutex<DeviceHandles>>>,
    Json(_body): Json<ChannelConfig>,
) -> Response {
    StatusCode::OK.into_response()
}

```

**Comment by bastibl** [comment]
cool project and thanks for the changes!

**Pull request merged**